IdentityCore/src/cache/key/MSIDMacKeychainNonshareableCredentialCacheKey.m

```diff
@@ -0,0 +1,130 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDMacKeychainNonshareableCredentialCacheKey.h"
+
+@implementation MSIDMacKeychainNonshareableCredentialCacheKey
+
+- (instancetype)initForRefreshTokenWithHomeAccountId:(NSString *)homeAccountId
+                                         environment:(NSString *)environment
+                                            clientId:(NSString *)clientId
+{
+    self = [super init];
+    
+    if (self)
+    {
+        _homeAccountId = homeAccountId;
+        _environment = environment;
+        _clientId = clientId;
+        _type = [NSNumber numberWithInt:2002];
+    }
+    
+    return self;
+}
+
+- (instancetype)initForAccessTokenWithHomeAccountId:(NSString *)homeAccountId
+                                        environment:(NSString *)environment
+                                              realm:(NSString *)realm
+                                           clientId:(NSString *)clientId
+                                             target:(NSString *)target
+{
+    self = [super init];
+    
+    if (self)
+    {
+        _homeAccountId = homeAccountId;
+        _environment = environment;
+        _realm = realm;
+        _clientId = clientId;
+        _type = [NSNumber numberWithInt:2001];
+        _target = target;
+    }
+    
+    return self;
+}
+
+- (instancetype)initForIdTokenWithHomeAccountId:(NSString *)homeAccountId
+                                    environment:(NSString *)environment
+                                          realm:(NSString *)realm
+                                       clientId:(NSString *)clientId
+{
+    self = [super init];
+    
+    if (self)
+    {
+        _homeAccountId = homeAccountId;
+        _environment = environment;
+        _realm = realm;
+        _clientId = clientId;
+        _type = [NSNumber numberWithInt:2003];
+    }
+    
+    return self;
+}
+
+- (NSString *)accountId
+{
+    NSString * normalizedHomeAccountId = self.homeAccountId.msidTrimmedString.lowercaseString;
+    NSString * normalizedEnvironment = self.environment.msidTrimmedString.lowercaseString;
+    return [NSString stringWithFormat:@"%@-%@", normalizedHomeAccountId, normalizedEnvironment];
+}
+
+- (NSString *)credentialId
+{
+    NSString * normalizedClientId = self.clientId.msidTrimmedString.lowercaseString;
+    NSString * normalizedRealm = self.realm.msidTrimmedString.lowercaseString;
+    return [NSString stringWithFormat:@"%@-%@-%@", self.type, normalizedClientId, normalizedRealm];
+}
+
+- (NSString *)account
+{
+    NSString* accessGroup = @"valentins_wonderful_access_group"; // TODO
+    NSString* appBundleId = [[NSBundle mainBundle] bundleIdentifier];
+    NSString* accountId = self.accountId;
+
+    return [NSString stringWithFormat:@"%@-%@-%@", accessGroup, appBundleId, accountId];
+}
+
+- (NSString *)service
+{
+    if(self.target.length > 0)
```

Comment:
This check shouldn't be necessary, as we still keep delimiters even if target is missing. Please see implementation in MSIDDefaultCredentialCacheKey:

```
// kSecAttrService - (<credential_type>-<client_id>-<realm>-<enrollment_id>-<target>)
- (NSString *)serviceWithType:(MSIDCredentialType)type
                     clientID:(NSString *)clientId
                        realm:(NSString *)realm
                 enrollmentId:(NSString *)enrollmentId
                       target:(NSString *)target
                       appKey:(NSString *)appKey
{
    realm = realm.msidTrimmedString.lowercaseString;
    clientId = clientId.msidTrimmedString.lowercaseString;
    target = target.msidTrimmedString.lowercaseString;
    enrollmentId = enrollmentId.msidTrimmedString.lowercaseString;

    NSString *credentialId = [self credentialIdWithType:type clientId:clientId realm:realm enrollmentId:enrollmentId];
    NSString *service = [NSString stringWithFormat:@"%@%@%@",
                         credentialId,
                         keyDelimiter,
                         (target ? target : @"")];
    
    if (![NSString msidIsStringNilOrBlank:appKey])
    {
        service  = [NSString stringWithFormat:@"%@|%@", service, appKey];
    }
    
    return service;
}
```

For that matter, perhaps we could even reuse "appKey" used for broker for macOS and update macOS spec (let's discuss)